File: javascript/DayTwelve.js
const fs = require('fs');

const WORKING = '.';
const DAMAGED = '#';
const UNKNOWN = '?';

function getInput(filePath) {
    return fs.readFileSync(filePath, 'utf8').split('\n').filter(Boolean);
}

function handle() {
    const input = getInput('../input/twelve.txt');
    const numberCounts = {};

    for (const line of input) {
        const unknownCount = (line.match(new RegExp('\\?', 'g')) || []).length;
        numberCounts[unknownCount] = (numberCounts[unknownCount] || 0) + 1;
    }

    for (const [n, count] of Object.entries(numberCounts)) {
        console.log(`${n} occurrences: ${count}`);
    }

    return handleStepOneDifferently();
}

function handleStepOneDifferently() {
    const input = getInput('../input/twelve.txt');
    const unknownPermDictionary = {};

    let totalCount = 0;
    for (let i = 0; i < input.length; i++) {
        const line = input[i];
        const [validConfigs, comboToCheck, localDict] = getValidConfigurations(line, unknownPermDictionary);

        for (const [key, value] of Object.entries(localDict)) {
            unknownPermDictionary[key] = value;
        }

        let lineCount = 0;
        for (const config of validConfigs) {
            if (doesLineProduceCorrectCombo(config, comboToCheck)) {
                lineCount++;
            }
        }

        console.log(`${line} ${lineCount}`);
        totalCount += lineCount;
    }

    return totalCount;
}


function getValidConfigurations(fullInputLine, unknownPermDictionary) {
    const validConfigs = new Set();
    const unknownPermutations = new Set();
    const [data, result] = fullInputLine.split(' ');
    const numberOfUnknowns = (data.match(new RegExp('\\?', 'g')) || []).length;
    const resultNumber = parseInt(result.replace(/,/g, ''));
    const unknownIndices = [...data].reduce((indices, c, i) => (c === UNKNOWN ? indices.concat(i) : indices), []);

    for (let i = 0; i <= numberOfUnknowns; i++) {
        const working = Array(i).fill(WORKING);
        const damaged = Array(numberOfUnknowns - i).fill(DAMAGED);
        unknownPermutations.add(working.join('') + damaged.join(''));
    }
    console.log({ unknownPermutations });

    const checkedPermWords = new Set();
    for (const perm of unknownPermutations) {
        const permPerms = unknownPermDictionary[perm] || permutations([...perm]);

        for (const newPerm of permPerms) {

            const permWord = newPerm;

            if (checkedPermWords.has(permWord)) {
                console.log("cacheHit");
                continue;
            }
            checkedPermWords.add(permWord);
            let newWord = data;

            for (let i = 0; i < numberOfUnknowns; i++) {
                newWord.replace('?', permWord[i]);
            }

            validConfigs.add(newWord);
        }
    }

    return [validConfigs, resultNumber, unknownPermDictionary];
}

function doesLineProduceCorrectCombo(lineData, comboToCheck) {
    let isCorrectCombo;
    let resultInt = 0;
    let brokenStringCount = 0;

    for (const char of lineData) {
        if (char === DAMAGED) {
            brokenStringCount++;
        } else if (brokenStringCount !== 0) {
            resultInt = resultInt * 10 + brokenStringCount;
            brokenStringCount = 0;
        }
    }

    if (brokenStringCount !== 0) {
        resultInt = resultInt * 10 + brokenStringCount;
    }

    isCorrectCombo = resultInt === comboToCheck;
    return isCorrectCombo;
}

if (require.main === module) {
    handle();
}

// Helper function for permutations
function* permutations(arr) {
    const length = arr.length;

    if (length <= 1) {
        yield [...arr];
        return;
    }

    for (let i = 0; i < length; i++) {
        const remaining = [...arr.slice(0, i), ...arr.slice(i + 1)];
        for (const p of permutations(remaining)) {
            yield [arr[i], ...p];
        }
    }
}

function* permute(input) {
    var i, ch;
    for (i = 0; i < input.length; i++) {
        ch = input.splice(i, 1)[0];
        usedChars.push(ch);
        if (input.length == 0) {
            permArr.push(usedChars.slice());
        }
        permute(input);
        input.splice(i, 0, ch);
        usedChars.pop();
    }
    return permArr
};

function onlyUnique(value, index, array) {
    return array.indexOf(value) === index;
}

// Helper function to add escape to RegExp (for older versions of Node.js)
RegExp.escape = function (s) {
    return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
};

File: Days/DayTwelve.cs
using System.Text.RegularExpressions;
using Utilities;

namespace Days;

public static class DayTwelve
{
    private static readonly char WORKING = '.';
    private static readonly char DAMAGED = '#';
    private static readonly char UNKNOWN = '?';
    private static readonly Regex UNKNOWN_REGEX = new Regex(Regex.Escape(UNKNOWN + ""));

    public static long Handle()
    {
        return HandleStepOneDifferently();
    }

    public static long HandleStepOneDifferently()
    {
        var input = InputReader.Get(".\\input\\twelve.txt");

        Console.WriteLine(input.Where(line => !line.Contains(UNKNOWN)).Count());

        var unknownPermDectionary = new Dictionary<string, List<string>>();

        // Parallel.For(10, 18, i =>
        // {
        //     Console.WriteLine("starting " + i);
        //     var dict = GeneratePreCacheData(i);
        //     foreach (var perm in dict)
        //     {
        //         lock (unknownPermDectionary)
        //         {
        //             unknownPermDectionary.TryAdd(perm.Key, perm.Value);
        //         }
        //     }
        //     Console.WriteLine("done " + i);
        // });

        long totalCount = 0;
        Parallel.For(0, input.Length, i =>
        {
            var line = input[i];
            var lineCount = 0;
            (var validConfigs, var comboToCheck, var localDict) =
            GetValidConfigurations(line, unknownPermDectionary);

            var localDictTest = localDict;
            foreach (var perm in localDictTest)
            {
                lock (unknownPermDectionary)
                {
                    unknownPermDectionary.TryAdd(perm.Key, perm.Value);
                }
            }

            foreach (var config in validConfigs)
            {
                if (DoesLineProduceCorrectCombo(config, comboToCheck))
                {
                    lineCount++;
                }
            }
            Console.WriteLine(line + " " + lineCount);
            totalCount += lineCount;
        });

        return totalCount;
    }

    private static Dictionary<string, IEnumerable<IEnumerable<char>>> GeneratePreCacheData(int numberOfQuestionMarks)
    {
        var unknownPermutations = new HashSet<string>();
        var unknownPermDectionary = new Dictionary<string, IEnumerable<IEnumerable<char>>>();

        for (int i = 0; i <= numberOfQuestionMarks; i++)
        {
            var working = new char[i];
            for (var j = 0; j < working.Length; j++)
            {
                working[j] = WORKING;
            }
            var damaged = new char[numberOfQuestionMarks - i];
            for (var j = 0; j < damaged.Length; j++)
            {
                damaged[j] = DAMAGED;
            }
            unknownPermutations.Add(string.Concat(working) + string.Concat(damaged));
        }

        foreach (var perm in unknownPermutations)
        {
            var permPerms = perm.Permutations();
            unknownPermDectionary.Add(perm, permPerms);
        }

        return unknownPermDectionary;
    }

    // expecting #....#... 1,1
    private static (HashSet<string>, long, Dictionary<string, List<string>>) GetValidConfigurations(string fullInputLine, Dictionary<string, List<string>> unknownPermDectionary)
    {
        var validConfigs = new HashSet<string>();
        var unknownPermutations = new HashSet<string>();
        var data = fullInputLine.Split(" ").First();
        var numberOfUnknowns = data.Count(a => a.Equals(UNKNOWN));
        var result = fullInputLine.Split(" ").Last();
        var numberOfResults = result.Split(",").Length;
        var resultNumber = int.Parse(result.Replace(",", ""));
        var unknownIndices = new List<int>();

        if (numberOfUnknowns == 0)
        {
            return (new HashSet<string>() { data }, resultNumber, unknownPermDectionary);
        }

        for (int i = 0; i <= numberOfUnknowns; i++)
        {
            var working = new char[i];
            for (var j = 0; j < working.Length; j++)
            {
                working[j] = WORKING;
            }
            var damaged = new char[numberOfUnknowns - i];
            for (var j = 0; j < damaged.Length; j++)
            {
                damaged[j] = DAMAGED;
            }
            unknownPermutations.Add(string.Concat(working) + string.Concat(damaged));
        }

        var checkedPermWords = new HashSet<string>();
        foreach (var perm in unknownPermutations)
        {
            var permPerms = perm.GetDistinctPermutations();
            // if (!unknownPermDectionary.TryGetValue(perm, out var permPerms))
            // {
            //     // permPerms = perm.Permutations();
            //     lock (unknownPermDectionary)
            //     {
            //         unknownPermDectionary.TryAdd(perm, permPerms);
            //     }
            // }

            foreach (var newPerm in permPerms)
            {
                var permWord = newPerm;

                if (checkedPermWords.TryGetValue(permWord, out var _))
                {
                    continue;
                }
                checkedPermWords.Add(permWord);
                var newWord = data;

                for (int i = 0; i < numberOfUnknowns; i++)
                {
                    newWord = UNKNOWN_REGEX.Replace(newWord, permWord[i] + "", 1, i);
                }
                validConfigs.Add(newWord);
            }
        }


        return (validConfigs, resultNumber, unknownPermDectionary);
    }

    private static bool DoesLineProduceCorrectCombo(string lineData, long comboToCheck)
    {

        bool isCorrectCombo;
        var resultInt = 0;
        var brokenStringCount = 0;

        for (int i = 0; i < lineData.Length; i++)
        {
            if (lineData[i] == DAMAGED)
            {
                brokenStringCount++;
            }
            else
            {
                if (brokenStringCount != 0)
                {
                    resultInt = resultInt * 10 + brokenStringCount;
                    brokenStringCount = 0;
                }
            }
        }

        if (brokenStringCount != 0)
        {
            resultInt = resultInt * 10 + brokenStringCount;
        }

        isCorrectCombo = resultInt == comboToCheck;
        return isCorrectCombo;
    }
}